network_simulator/host.py

```diff
@@ -23,4 +28,91 @@ def __init__(self, address=None, flows={}, flow_packets_received={},
         self.flow_packets_received = flow_packets_received
         self.link = link
 
-    # TODO(team): More event subclasses to come?
\ No newline at end of file
+
+class HostReceivedPacketEvent(Event):
+    """
+    Host receives packets from a flow.
+    """
+
+    def __init__(self, host, packet, packet_previously_received):
+        """
+        Sanity check: make sure destination was this Host and that this isn't
+        a RouterPacket.
+        :ivar host: Host that receives Packet
+        :ivar packet: Packet received
+        :ivar packet_previously_received: True if Packet previously received
+        """
+        Event.__init__(self)
+
+        # Sanity check
+        assert packet.dest_id == host.address and not isinstance(packet,
+                                                                 RouterPacket)
+        self.host = host
+        self.packet = packet
+        self.packet_previously_received = packet_previously_received
+
+    def elem_in_list(elem, my_list):
+        # ind is where this element *would* be inserted
+        # into the list in order to keep it in sorted order.
+        ind = bisect.bisect_left(my_list, elem)
+        if ind != len(my_list) and my_list[ind] == elem:
+            return True
+        else:
+            return False
+
+    def run(self, main_event_loop, statistics):
+        """
+        Put packet in flow_packets_received and sort.
+        Update statistics.
+        :param main_event_loop: event loop where new Events will be scheduled.
+        :param statistics: statistics to update
+        """
+        # add packet to host.flow_packets_received
+        old_packets_received = self.host.flow_packets_received.get(
```

Comment:
All of the logic about old_packets_received and the if statement below has to be surrounded in another if, since you need to check if isinstance(packet, DataPacket). We are only looking at DataPackets received by this Host.